awx/ui_next/src/screens/Project/ProjectList/useWsProjects.js

```diff
@@ -0,0 +1,85 @@
+import { useState, useEffect, useRef } from 'react';
+
+export default function useWsProjects(initialProjects) {
+  const [projects, setProjects] = useState(initialProjects);
+  const [lastMessage, setLastMessage] = useState(null);
+  const ws = useRef(null);
+
+  useEffect(() => {
+    setProjects(initialProjects);
+  }, [initialProjects]);
+
+  useEffect(() => {
+    if (!lastMessage?.unified_job_id || lastMessage.type !== 'project_update') {
+      return;
+    }
+    const index = projects.findIndex(p => p.id === lastMessage.project_id);
+    if (index === -1) {
+      return;
+    }
+
+    const project = projects[index];
+    const updatedProject = {
+      ...project,
+      summary_fields: {
+        ...project.summary_fields,
+        last_job: {
+          id: lastMessage.unified_job_id,
+          status: lastMessage.status,
+          finished: lastMessage.finished,
+        },
+      },
+    };
+    setProjects([
+      ...projects.slice(0, index),
+      updatedProject,
+      ...projects.slice(index + 1),
+    ]);
+  }, [lastMessage]); // eslint-disable-line react-hooks/exhaustive-deps
```

Comment:
Do we need to monitor `lastMessage` with `useEffect` ? Adding `useEffect` seems like it is adding an unnecessary second step to event processing. What about putting the project updating logic in a function and triggering it directly from the events emitted by the websocket?

Reply:
This approach is based on one I'd seen in some examples online with WS and React hooks. I want to say there was a reason it didn't work the other way, but I'll verify that...